File: bro-agent-api/src/main/cpp/service/RestoreService.cc
#include <cstdio>

#include "exception/FailedToDownloadException.hh"

#include "RestoreService.hh"

using namespace com::ericsson::adp::mgmt::data;

namespace BackupRestoreAgent
{

RestoreService::RestoreService(RestoreDataStream restoreStream, IStream* outStream)
    : restoreStream(std::move(restoreStream))
    , outStream(outStream)
{
}

RestoreService::~RestoreService()
{
    delete outStream;
    grpc::Status status = restoreStream->Finish();
    if (!status.ok())
    {
        std::cerr << "Cannot finish reading RestoreData stream!" << std::endl;
    }
}

void RestoreService::download(const std::string& restoreLocation)
{
    std::cerr << "RestoreService::download" << std::endl;
    try
    {
        RestoreData server_data;
        while (restoreStream->Read(&server_data))
        {
            parseServerData(server_data, restoreLocation);
        }
    }
    catch (...)
    {
        throw new FailedToDownloadException("Error at Orchestrator while downloading data during restore");
    }
}

void RestoreService::parseServerData(const RestoreData& server_data, const std::string& restoreLocation)
{
    switch (server_data.datamessagetype())
    {
    case DataMessageType::CUSTOM_METADATA_FILE:
        parseFileData(server_data, &RestoreData::custommetadatafilechunk, restoreLocation);
        break;
    case DataMessageType::BACKUP_FILE:
        parseFileData(server_data, &RestoreData::backupfilechunk, restoreLocation);
        break;
    default:
        break;
    }
}

template <class T, class FType>
void RestoreService::parseFileData(const T& server_data, FType getfilechunk, const std::string& restoreLocation)
{
    if ((server_data.*getfilechunk)().filename() != "")
    {
        restoreFile = restoreLocation + "/" + (server_data.*getfilechunk)().filename();
        outStream->open(restoreFile);
    }
    if ((server_data.*getfilechunk)().content() != "")
    {
        outStream->write((server_data.*getfilechunk)().content());
        calculator.addBytes((server_data.*getfilechunk)().content());
    }
    if ((server_data.*getfilechunk)().checksum() != "")
    {
        outStream->close();
        std::string checksum = calculator.getChecksum();
        if (checksum.compare((server_data.*getfilechunk)().checksum()) != 0)
        {
            std::cerr << "CHECKSUM mismatch: " << checksum << std::endl;
            std::remove(restoreFile.c_str());
        }
        calculator = ChecksumCalculator();
    }
}
}

File: bro-agent-api/src/main/python/bro_agent/service/RestoreService.py
import logging

import bro_agent.generated.INT_BR_ORCH_DATA_pb2 as INT_BR_ORCH_DATA
from bro_agent.util.ChecksumCalculator import ChecksumCalculator


class RestoreService:
    def __init__(self, restore_location):
        self.calculator = ChecksumCalculator()
        self.restore_location = restore_location

    def download(self, restore_data_iterator):
        for data in restore_data_iterator:
            if data.dataMessageType is INT_BR_ORCH_DATA.DataMessageType.BACKUP_FILE:
                self.write_backup_file(data.backupFileChunk)
            if data.dataMessageType is INT_BR_ORCH_DATA.DataMessageType.CUSTOM_METADATA_FILE:
                self.write_custom_metadata_file(data.customMetadataFileChunk)
        logging.info("Finished downloading data!")

    def write_backup_file(self, chunk):
        if chunk.fileName != "":
            self.fileName = chunk.fileName
            self.fileDescriptor = open(self.restore_location + '/' + self.fileName, "wb")
        elif chunk.content != b"":
            self.fileDescriptor.write(chunk.content)
            self.calculator.add_bytes(chunk.content)
        elif chunk.checksum != b"":
            self.fileDescriptor.close()
            if chunk.checksum == self.calculator.get_checksum().upper():
                logging.info("Checksum matched!")
            else:
                logging.error("Checksum mismatch!")

    def write_custom_metadata_file(self, chunk):
        self.write_backup_file(chunk)
